### ISSUE TITLE
In program updates incompatible with "Child process" mitigations

### ISSUE THREAD

**Issue by rugabunda** [telegramdesktop/tdesktop#16627]
Never had this issue before now, had CIG enabled in telegram for months, updated it many times without issue.

![CIG+Telegram 2 8 8](https://user-images.githubusercontent.com/38260267/125872407-f061f9bd-510b-4956-8f5e-bba1d14aa218.png)

Telegram 2.8.8 fails to load after installed. Error messages appear on screen. Stating "bad image".

Steps to recreate: Enable Windows 10 security mitigations, enable allow only images signed by microsoft...  then try run or install Telegram 2.8.8. May need winscp, openshell & adobe products installed such as photoshop & premiere.

```
Process '\Device\HarddiskVolume4\Users\Ty\AppData\Roaming\Telegram Desktop\Telegram.exe' (PID 17360) was blocked from loading the non-Microsoft-signed binary '\Program Files (x86)\WinSCP\DragExt64.dll'.

Process '\Device\HarddiskVolume4\Users\Ty\AppData\Roaming\Telegram Desktop\Telegram.exe' (PID 17360) was blocked from loading the non-Microsoft-signed binary '\Windows\System32\StartMenuHelper64.dll'.

Process '\Device\HarddiskVolume4\Users\Ty\AppData\Roaming\Telegram Desktop\Telegram.exe' (PID 17360) was blocked from loading the non-Microsoft-signed binary '\Program Files (x86)\Common Files\Adobe\CoreSyncExtension\CoreSync_x64.dll'.

Process '\Device\HarddiskVolume4\Users\Ty\AppData\Roaming\Telegram Desktop\Telegram.exe' (PID 17360) was blocked from loading the non-Microsoft-signed binary '\Program Files\Open-Shell\ClassicExplorer64.dll'.

```

Strange. These are non malicious DLLS. Winscp is not even loaded. Coresync & Openshell is. Never saw this before 2.8.8. Is your update supply chain in tact?

**Comment by rugabunda** [comment]
Here is a video of the error, right after downloading 2.8.8 from within Telegram. https://vimeo.com/575610675

**Comment by tcNmjeD6SpVrpLC** [comment]
How is this related to telegram? You use drag and drop feature by winscp, don't you?

**Comment by john-preston** [comment]
@rugabunda Now tdesktop is forced to load libraries only from system locations and maybe I try to load some libraries I shouldn't have (as part of protection from dll preloading attacks before I tried to load many dlls from system folders only). I'll try making the list smaller in the next versions.

**Comment by rugabunda** [comment]
> 
> 
> Here is a video of the error, right after downloading 2.8.8 from within Telegram. https://vimeo.com/575610675

Ah yes I see in Autoruns, winscp Drag&Drop shell extension is loaded on boot as a Shellex CopyHook Handle.

Tx John.

**Comment by rugabunda** [comment]
> 
> 
> @rugabunda Now tdesktop is forced to load libraries only from system locations and maybe I try to load some libraries I shouldn't have (as part of protection from dll preloading attacks before I tried to load many dlls from system folders only). I'll try making the list smaller in the next versions.

Can you provide the list you are allowing Telegram to inject into its process? And why is telegram loading unnecessary and unrelated dlls?  It is loading unsigned binaries like "ClassicExplorer64.dll" into telegram.

**Comment by john-preston** [comment]
@rugabunda Actually, this is strange, because I stopped loading a lot of .dlls (which were loaded before from System32 to prevent dll preloading from the folder containing the Telegram.exe). Now it just enforces loading only from System32:

https://github.com/telegramdesktop/tdesktop/commit/148af59615aac7e5b31ca15ba7deece7e0fee347#diff-0d41a79992947d35ab801c35d32f67f1052487ab9663f2ae67cd7da19f638a79

https://github.com/desktop-app/lib_base/blob/f9ca7dc7f357ecafb9ae9b5ffd467f364b6b124e/base/platform/win/base_windows_safe_library.cpp#L80-L90

So I don't know why they're being loaded.

**Comment by tcNmjeD6SpVrpLC** [comment]
Because on windows 10 you can't just ignore those dll's, most of them related to drag and drop features due authority level of explorer.exe processes, too complex to describe it here.

Also you can change the preferable order of loading libraries by using flags in LoadLibrary (Ex).

**Comment by john-preston** [comment]
I'm not sure I understand. What do you mean I can't ignore those dll's? As far as I can tell, I do nothing with any dll's right now. I just allow the loading only from system directory by calling `SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32);`

**Comment by tcNmjeD6SpVrpLC** [comment]
Answer mainly related to https://github.com/telegramdesktop/tdesktop/issues/16627#issuecomment-881375983. The second part for you, correct.

**Comment by john-preston** [comment]
But why should I change order of loading libraries? And which libraries? I load only the system libraries that I need, here:

https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/SourceFiles/platform/win/windows_dlls.cpp#L28-L71

and some others.

**Comment by ilya-fedin** [comment]
>  It is loading unsigned binaries like "ClassicExplorer64.dll" into telegram.

All what's changed is that system libraries are linked with DELAYLOAD now and `SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32);` is used, like @john-preston. Whatever injects ClassicExplorer64.dll for you, it's not tdesktop itself.
https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/CMakeLists.txt#L1398-L1422

**Comment by rugabunda** [comment]
> 
> 
> > It is loading unsigned binaries like "ClassicExplorer64.dll" into telegram.
> 
> All what's changed is that system libraries are linked with DELAYLOAD now and `SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32);` is used, like @john-preston mentioned. Whatever injects ClassicExplorer64.dll for you, it's not tdesktop itself.
> https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/CMakeLists.txt#L1398-L1422

So you and I can install 2.8.3 no problem with CIG mitigations... its working fine. We can recreate this problem every time using 2.8.8. It appears DELAYLOAD & SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32) are allowing third party dll's outside of system32 to be injected into Telegram.

**Comment by ilya-fedin** [comment]
> So you and I can install 2.8.3 no problem with CIG mitigations...

I think your program just is not really good at injection detection and skips libraries heavy linked to binaries. As you can see, the change doesn't adds libraries you mentioned from Telegram, so the only logical conclusion is that these libraries are injected by system libraries (via some shell integrations)  and your software wasn't good enough to detect that prior to 2.8.8.

**Comment by rugabunda** [comment]
> 
> 
> > So you and I can install 2.8.3 no problem with CIG mitigations...
> 
> I think your program is just not really good at injection detection and skips libraries heavy linked to binaries. As you can see, the change doesn't add libraries you mentioned to Telegram, so the only logical conclusion is that these libraries are injected by system libraries (via some shell integrations) and your software wasn't good enough to detect that prior to 2.8.8.

So you are saying microsoft's built in CIG mitigations did not work until Telegram utilized (SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32). However, I have looked don't see any of the aforementioned libraries loaded into Telegram in version 2.8.3.

**Comment by ilya-fedin** [comment]
> So you are saying microsoft's built in CIG mitigations did not work until Telegram utilized (SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32).

I don't think this call has to do with that, more likely it's that transform from the common link to DELAYLOAD (MSVC transforms this to runtime code with LoadLibrary, so that these loads are guarded by SetDefaultDllDirectories). Apparently, your software hooks only LoadLibrary and doesn't work whren there's no LoadLibrary in the middle.

**Comment by ilya-fedin** [comment]
> However, I have looked and none of the aforementioned libraries loaded into Telegram in version 2.8.3. I can further test by disabling CIG and looking at 2.8.8 as well.

You can check the source code and you will see that none off these libraries are requested by Telegram. The problem is somewhere in your system and the new library load method just showed that to you.

**Comment by rugabunda** [comment]
Actually, taking a look at Event Viewer, I see that CIG is and was already blocking third part libraries from being injected into Telegram in version 2.8.3, except it was not causing the program to crash. 2.8.8 causes the program to actually crash. These warnings only began appearing on 7/7/2021. 

The latest update for Telegram also breaks another windows security feature, "do not allow child process."

Process '\Device\HarddiskVolume4\Users\Ty\AppData\Roaming\Telegram Desktop\Telegram.exe' (PID 7248) was blocked from creating a child process 'C:\Users\Ty\AppData\Roaming\Telegram Desktop\Updater.exe' with command line '"C:\Users\Ty\AppData\Roaming\Telegram Desktop\Updater.exe" -update -exename "Telegram.exe"'.

 but the latest update breaks this security feature. I just updated to 21H1 as well, I wonder if they enhanced some of the mitigations.

**Comment by ilya-fedin** [comment]
>  see that CIG is and was already blocking third part libraries from being injected into Telegram in version 2.8.3

Still, these libraries aren't requested by Telegram, they're loaded by system libraries, there's nothing Telegram can about that, you installed software that hooks in your system.

> The latest update for Telegram also breaks another windows security feature mitigation, "do not allow child process." For updates. Never had this problem before either.

Telegram always had Updater binary and called it for updates. Nothing was changed about that except of that Updater.exe uses DELAYLOAD as well.

**Comment by rugabunda** [comment]
> , there's nothing Telegram can do about that, you've installed software that hooks your system.

Whatever telegram did about it worked fine, before 2.8.8. The changes in 2.8.8 created this problem.

**Comment by ilya-fedin** [comment]
> Whatever telegram did about it worked fine, before 2.8.8. The changes in 2.8.8 created this problem.

This doesn't mean they should be reverted just because of your problems with your system

**Comment by rugabunda** [comment]
Doesn't mean the shouldn't, this needs to be looked at closely.

**Comment by ilya-fedin** [comment]
And looks like only you have this problem

**Comment by rugabunda** [comment]
Why do I find that nearly impossible to believe? I will test in virtual machines.

**Comment by ilya-fedin** [comment]
Because you don't understand probably that these libraries are loaded by your system libraries, not by Telegram itself

**Comment by rugabunda** [comment]
I understand that, but 2.8.8 breaks this mitigation in windows, making telegram more vulnerable to exploitation.

**Comment by ilya-fedin** [comment]
> I understand that, but 2.8.8 breaks this mitigation in windows, making telegram more vulnerable to exploitation.

What do you mean by that? Telegram still loads the same system libraries. If one of your system libraries are hooked, Telegram is no longer responsible for that

**Comment by rugabunda** [comment]
Because the only way to load telegram now is to disable CIG mitigations for telegram

**Comment by ilya-fedin** [comment]
> Because the only way to load telegram now is to disable CIG mitigations for telegram

This means that this 'CIG mitigations' (whatever it is) is just skipped these hooks previously.

**Comment by ilya-fedin** [comment]
> telegram is no longer allowing this protection

I see on your screenshot that it's blocked something, so it allowed to protect rather than denied

**Comment by rugabunda** [comment]
> 
> 
> > Because the only way to load telegram now is to disable CIG mitigations for telegram
> 
> This means that this 'CIG mitigations' (whatever it is) just skipped these hooks previously.

Please explain how you came to the conclusion that windows security skipped this before, and how changes in telegram 2.8.8 suddenly it began working?

**Comment by ilya-fedin** [comment]
> Please explain, if you are capable, how you came to the conclusion that windows security skipped this before, and how changes in telegram 2.8.8 suddenly it began working?

You say that it allowed to run the executable previously and it doesn't now. Since tdesktop never used these libraries, the only logical conclusion is that it skipped these hooks previously.

**Comment by rugabunda** [comment]
Explain to me how I can determine this with certainty. Perhaps **SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32)** has documented or undocumented affect on microsofts own Code Integrity Guard; or perhaps CIG it makes SetDefaultDllDirectories  more effective?

The coder responsible already said  "Now tdesktop is forced to load libraries only from system locations and **maybe I try to load some libraries I shouldn't have** (as part of protection from dll preloading attacks before I tried to load many dlls from system folders only). **I'll try making the list smaller in the next versions.**"

**Comment by ilya-fedin** [comment]
> Explain to me how I can determine this with certainty.

I would honestly say I have no idea, I'm not a security expert.

> Perhaps **SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32)** has documented or undocumented affect on microsofts own Code Integrity Guard; or perhaps CIG it makes SetDefaultDllDirectories more effective?

Why are you talking about it, I said previously that it's unlikely that SetDefaultDllDirectories has to do anything with that

> **maybe I try to load some libraries I shouldn't have**

I think he meant he may loading system libraries that aren't really needed (only system libraries are loaded, you can see that list by the link I sent in one of previous messages), but still these libraries are hooked in by system libraries, not by tdesktop.

> No other application in windows is receiving these errors.

So this 'CIG mitigations' is apparently of very low quality if it doesn't warn about these hooks for other applications.

**Comment by rugabunda** [comment]
It works, but no other program loads these DLLs except for explorer. These hooks are loaded into explorer.exe, nowhere else. If I enable CIG mitigations on explorer.exe, explorer fails to load, windows wont boot.

**Comment by ilya-fedin** [comment]
> It works, but no other program loads these DLLs except for explorer.

Have you tested other Qt applications?

**Comment by rugabunda** [comment]
Yes. I tried it on Telegram, last week, it worked fine.

**Comment by rugabunda** [comment]
Name another QT program

**Comment by ilya-fedin** [comment]
> Yes. I tried it on Telegram, last week, it worked fine.

The only library that was added with that time is ANGLE, Google's library for transforming GLES calls to DirectX

**Comment by ilya-fedin** [comment]
> Name another QT program

qBittorrent, but almost all other Qt programs don't use hardware rendering and ANGLE

**Comment by rugabunda** [comment]
I use qbittorrent for years, I will check if CIG is enabled.  I have disabled CIG probably due to incompatibility. I will re-test and report results shortly. CIG is hit and miss with many third party non microsoft applications. It only allows microsoft signed DLLs. When it actually works with third party software, that is a very good thing, because you can be sure nothing can inject third party dlls into your software.  SetDefaultDllDirectories is very similar to CIG;

**Found the following, searching for CIG & SetDefaultDllDirectories** 

**DLL Hijacking - Why running executables from a user-writable location is a bad idea.**
https://dennisbabkin.com/blog/?t=windows-security-legacy-dll-hijacking-running-executables-from-user-writable-location

```
Lastly, with the same restrictions imposed on the operating system (anything older than Windows 7, or you will need KB2533623 installed) you can use the following alternative method of specifying the search order globally with the SetDefaultDllDirectories function: 

But, as you can see, the amount of code involved to implement it is way more than just calling the following for the static linking:

So it is obvious that most developers, including Microsoft itself, will not be willing to go this route.

It seems like Microsoft had a good idea with their Code Integrity Guard, or CIG. It's only if they implemented it thoroughly.
```

https://dennisbabkin.com/blog/?t=windows-security-legacy-dll-hijacking-running-executables-from-user-writable-location

This demonstrates how TrendMIcro Maximum Security & TrendMicro security software was hijacked and implanted with persistent malware because the CIG (Code Integrity Guard) mechanism is not enforced; they also suggest using SetDefaultDllDirectories as a mitigation.

https://www.safebreach.com/blog/2019/trend-micro-security-16-dll-search-order-hijacking-and-potential-abuses/
https://www.safebreach.com/blog/2019/BitDefender-Antivirus-Free-2020/

`Despite the fact that it’s an antivirus, this service run as a non-PPL, which means that CIG (Code Integrity Guard) is not enforced, so unsigned code-loading is possible into these processes.
`

**Comment by ilya-fedin** [comment]
How much times I should repeat that it's unlikely that SetDefaultDllDirectories is the cause?

**Comment by rugabunda** [comment]
Say it as much as you like. You said only two things as changed, SetDefaultDllDirectories & DELAYLOAD  are the only changes.
> 
> > It is loading unsigned binaries like "ClassicExplorer64.dll" into telegram.
> 
> All what's changed is that system libraries are linked with DELAYLOAD now and `SetDefaultDllDirectories(LOAD_LIBRARY_SEARCH_SYSTEM32);` is used, like @john-preston mentioned. Whatever injects ClassicExplorer64.dll for you, it's not tdesktop itself.
> https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/CMakeLists.txt#L1398-L1422

**Comment by ilya-fedin** [comment]
> You said only two things as changed, SetDefaultDllDirectories & DELAYLOAD are the only changes.

And ANGLE was added

**Comment by rugabunda** [comment]
I just re-enabled CIG on Qbittorrent. So far, no problems at all. I see no third party dll's loading there, just as I saw no third party dlls loading in 2.8.3.

**Comment by ilya-fedin** [comment]
The thing is that You said only two things as changed, is just a call, while DELAYLOAD changes how libraries are loaded

**Comment by ilya-fedin** [comment]
> I just re-enabled CIG on Qbittorrent. So far, no problems at all. I see no third party dll's loading there, just as I saw no third party dlls loading in 2.8.3.

I still think that CIG just ignores hooking when DELAYLOAD is not used

**Comment by ilya-fedin** [comment]
You can try to disable ANGLE via `Settings > Advanced > Performance > ANGLE`, but I'm not sure whether this fully prevents calls to it or not, @john-preston should know better

**Comment by rugabunda** [comment]
I did from the moment it was offered, as I already had opengl disabled

**Comment by rugabunda** [comment]
I unsuccessfully tried recreating this issue in a virtual machine, however there was a new version already pushed. 2.8.10

**Comment by rugabunda** [comment]
Ok, after upgrading to 2.8.10 on my primary machine, this issue no longer persists.

But SetDefaultDllDirectories basically does what CIG does, through not specific to microsoft signed dlls. Meaning, SetDefaultDllDirectories should make CIG mostly unnecessary.

**Comment by rugabunda** [comment]
So I tried again in the VM. With "child process" mitigations, I had no error popup window, which is a good sign. But I saw the following two in event viewer, in this order respectively:

```
Process '\Device\HarddiskVolume3\Users\pc\AppData\Roaming\Telegram Desktop\Telegram.exe' (PID 5776) was blocked from creating a child process 'C:\Users\pc\AppData\Roaming\Telegram Desktop\Updater.exe' with command line '"C:\Users\pc\AppData\Roaming\Telegram Desktop\Updater.exe" -update -exename "Telegram.exe"'.

Process '\Device\HarddiskVolume3\Users\pc\AppData\Roaming\Telegram Desktop\Telegram.exe' (PID 5776) was blocked from loading the non-Microsoft-signed binary '\Program Files\Classic Shell\ClassicExplorer64.dll'.
```

**Comment by rugabunda** [comment]
If you can provide a link to 2.8.8 installer? I would like to confirm some of my testing here just to be certain

**Comment by rugabunda** [comment]
Strange. After first installing 2.8.8, (directly from the downloaded installer) it installed and loaded initially without any errors. After closing it, the first time loading it manually, it pops up with the aforementioned "bad Image" errors, sometimes it does not. The third time loading it, it loads fine without any errors. This is with CIG and child process mitigations are in place.

**Comment by rugabunda** [comment]
OK i switched back to 2.8.6 beta, tried updating with child protections on, same problem I had initially... it would not install; even after closing & re-loading telegram after downloading the update inside the program, instead of clicking "UPDATE TELEGRAM", in hopes of preventing a "child process".

After disabling the child process mitigation, it installed and everything worked fine with CIG enabled.

**Comment by john-preston** [comment]
Child Process Mitigations forbids the software to create child processes?

If so, there is no intention to make it work with tdesktop, I have no idea how to make autoupdates without making child processes and in the future Google Crashpad for crash report writing may be used, that AFAIK is launched in a separate process.

**Comment by john-preston** [comment]
@rugabunda The changes made in the latest versions are exactly the ones described in OneDrive article mentioned above. Now all libraries are loaded dynamically, after the call to SetDefaultDllDirectories.

Those changes mitigate dll hijacking, which could be done before.

The part with "maybe I try to load some libraries I shouldn't have" and "I'll try making the list smaller in the next versions." was written before I checked and saw, that in the latest versions I stopped loading unnecessary dlls (which was a protection from dll hijacking before, because I tried loading them only from System32 folder) and now this list is empty anyway, I load only system libraries, required for the software to work.
